### PR TITLE
StyleNano removal part 1: New stylesheet system

### PR DESCRIPTION
**Pull request by moonheart08**
## About the PR
it (will be) gonb. Currently just staking a claim so people know what I'm doing.

## Why / Balance
https://github.com/space-wizards/space-station-14/blob/1cf07fadace1dd4d37967daf3e30155561712ad7/Content.Client/Stylesheets/StyleNano.cs#L19-L1485

## Technical details
This literally just rewrites content-side stylesheets from the ground up.
### Major API/design changes
- Hardcoding colors/textures/etc instead of using the stylesheet is now hard-and-fast illegal. All color constants are private, all utilities for reading textures for the stylesheet are access-locked, etc.
- The main stylesheet (StyleNano) is now palette-based, with most colors being derived from a set of primaries.
- Sheetlets have been introduced, as to remove the need to modify the main stylesheet to style specific UI elements.
- DefaultWindow should no longer be used in content. Please.

## Breaking changes
If you touched StyleNano, you're going to need to redo your stuff.

## Future work
- Allow the user to change the five palette colors (`primary-1`, `secondary-1`, `positive-1`, `negative-1`, `highlight-1`)
- Allow the user to change the primary font.
- Syndicate/domain specific stylesheets (because they're no longer a major maintenance hazard)
- Reorganization. At this time all new tools, sheets, code, etc are in Content.Client/Stylesheets/Redux for ease of review.

**Changelog**

:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
- tweak: Many UI colors have been fine-tuned (new palettes)

### CODE REVIEW

**Comment by moonheart08** [comment]
Preliminary documentation: https://github.com/space-wizards/space-station-14/blob/f411a5b6a9cc2cb71b280cdc2568f60c71c2942d/Content.Client/Stylesheets/Redux/Stylesheets.md

**Comment by moonheart08** [comment]
Reviewable, but not mergable.

**Comment by eoineoineoin** [comment]
Really liking the possibilities this is opening up - the MutableSelectorChild is great. One thing I think I'd like to be able to do is to select elements by their particular name. Suppose I have a control:

```
<MyWidget>
    <Label Name="ImportantLabel" />
    <Label Name="FlavourLabel" />
```

And I want to write a few different styles for MyWidget, where those styles would want different settings on the ImportantLabel Label. I'd like to write a selector rule like:

```new MutableSelectorChild().Parent(E<MyWidget>().Class("OMG")).Child(E<Label>().ElementNamed("ImportantLabel"))```

But (AFAICS) there's no selector for object names? Now, I could add a style class to the labels:

```
<MyWidget>
    <Label Name="ImportantLabel" StyleClasses="StyleImportant" />
    <Label Name="FlavourLabel" StyleClasses="StyleSubtle" />
```

And then the selector would become:

``````new MutableSelectorChild().Parent(E<MyWidget>().Class("OMG")).Child(E<Label>().Class("StyleImportant"))```

So it's not too bad (MenuButton has this kind of strategy), but feels like it's extra friction - the styleClass might only be used to ever identify that one element (no other widget going to use the Button's topButtonLabel style, for example) so it's essentially acting as the element name.

**Comment by eoineoineoin** [comment]
Also, I'm aware it's work in progress, but I'm finding the colour of the disabled buttons kind of jarring, as the low contrast between the background makes it hard to see the borders:

image

Compared with before:

image

**Comment by moonheart08** [comment]
> Really liking the possibilities this is opening up - the MutableSelectorChild is great. One thing I think I'd like to be able to do is to select elements by their particular name. Suppose I have a control:
> 
> ```
> <MyWidget>
>     <Label Name="ImportantLabel" />
>     <Label Name="FlavourLabel" />
> ```
> 
> And I want to write a few different styles for MyWidget, where those styles would want different settings on the ImportantLabel Label. I'd like to write a selector rule like:
> 
> `new MutableSelectorChild().Parent(E<MyWidget>().Class("OMG")).Child(E<Label>().ElementNamed("ImportantLabel"))`
> 
> But (AFAICS) there's no selector for object names? Now, I could add a style class to the labels:
> 
> ```
> <MyWidget>
>     <Label Name="ImportantLabel" StyleClasses="StyleImportant" />
>     <Label Name="FlavourLabel" StyleClasses="StyleSubtle" />
> ```
> 
> And then the selector would become:
> 
> `new MutableSelectorChild().Parent(E<MyWidget>().Class("OMG")).Child(E<Label>().Class("StyleImportant"))`
> 
> So it's not too bad (MenuButton has this kind of strategy), but feels like it's extra friction - the styleClass might only be used to ever identify that one element (no other widget going to use the Button's topButtonLabel style, for example) so it's essentially acting as the element name.

tl;dr "No."
We already have three levels of specificity for elements, ID, Class, and Type. We do not need a fourth, use styleclasses. Making the element's XAML name meaningful would be an antipattern in the face of refactors.

(also MutableSelectorChild has existed for years, this adds a helper for it, `x.ParentOf(y)`, and no more)

**Comment by moonheart08** [comment]
> Also, I'm aware it's work in progress, but I'm finding the colour of the disabled buttons kind of jarring, as the low contrast between the background makes it hard to see the borders:
> 
> image
> 
> Compared with before:
> 
> image

noted, it was set temporarily so i'll change it later.

**Comment by ElPidor** [comment]
SICK!
I hope uplink will look badass and giga ultra syndicate as in 13
